Increase max number of dimensions from 16 to 21

salvatore-campagna opened the pull request:
Here we increase the maximum number of dimension fields
for a time series index from 16 to 21. The maximum size of each.
The maximum allowed size for a field in Lucene is 32 Kb.

When encoding the tsid we include all dimension field names and
all dimension field values. As a result we have, max field name 512
bytes, max field value 1024 bytes. 32 Kb / (512 + 1024) bytes = 21.3.
So the maximum number of dimensions is 21.

See #93564

salvatore-campagna committed:
fix: increase max number of dimensions from 16 to 21

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

salvatore-campagna committed:
fix: checkstyle format violation

salvatore-campagna committed:
fix: tests using previous dimension fields limits

salvatore-campagna committed:
fix: always use 21 as dimension limit

salvatore-campagna committed:
chore: remove unnecessary braces

martijnvg commented on server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesIdFieldMapper.java:
```diff
@@ -67,7 +67,7 @@ public class TimeSeriesIdFieldMapper extends MetadataFieldMapper {
      * hundred bytes each, but we're being paranoid here.
      */
-    private static final int DIMENSION_VALUE_LIMIT = 1024;
+    private static final int DIMENSION_VALUE_LIMIT = 1536;
```
I don't think we can increase this constant? The limit per field is DIMENSION_VALUE_LIMIT + DIMENSION_NAME_LIMIT. Otherwise we may spend up to ~2kb per field and then we hit the 32kb limit with less then 21 fields?

salvatore-campagna replied:
Ok I missed that...the tsid includes both the name and the value.

salvatore-campagna committed:
fix: restore value limit to 1024

The tsid encodes both the field name and the field value. As a result,
1536 bytes is the sum of the field name max size and the field value
max size, respectively 512 and 1024.

salvatore-campagna committed:
fix: restore value limit to 1024

martijnvg commented on server/src/internalClusterTest/java/org/elasticsearch/timeseries/support/TimeSeriesDimensionsLimitIT.java:
```diff
@@ -0,0 +1,181 @@
+        int dimensionFieldLimit = 21;
+        final Exception ex = expectThrows(IllegalArgumentException.class, () -> createTimeSeriesIndex(mapping -> {
+            mapping.startObject("routing_field").field("type", "keyword").field("time_series_dimension", true).endObject();
+            for (int i = 0; i < dimensionFieldLimit; i++) {
```
Maybe I'm reading this incorrectly. This creates a mapping with 21 dimension fields. But this should succeed? Only if there are 22 dimensions this should fail?

salvatore-campagna replied:
There is the routing_field that is a dimension so it is actually 22. I need to include something in the routing path otherwise the test fails because there is no routing path.

martijnvg commented on server/src/internalClusterTest/java/org/elasticsearch/timeseries/support/TimeSeriesDimensionsLimitIT.java:
```diff
@@ -0,0 +1,181 @@
+            .putList(IndexMetadata.INDEX_ROUTING_PATH.getKey(), routingPaths.get())
+            .put(IndexSettings.TIME_SERIES_START_TIME.getKey(), "2000-01-08T23:40:53.384Z")
+            .put(IndexSettings.TIME_SERIES_END_TIME.getKey(), "2106-01-08T23:40:53.384Z")
+            .put(MapperService.INDEX_MAPPING_DIMENSION_FIELDS_LIMIT_SETTING.getKey(), dimensionsFieldLimit)
```
Now we always concrete set the limit. Maybe also test with the default?

martijnvg reviewed (commented):
Looks good. I left 2 questions.

salvatore-campagna committed:
fix: test default number of dimensions

salvatore-campagna committed:
Merge branch 'elastic:main' into feature/increase-dimensions-limit

martijnvg commented on server/src/internalClusterTest/java/org/elasticsearch/timeseries/support/TimeSeriesDimensionsLimitIT.java:
```diff
@@ -0,0 +1,200 @@
+            .put(IndexSettings.TIME_SERIES_START_TIME.getKey(), "2000-01-08T23:40:53.384Z")
+            .put(IndexSettings.TIME_SERIES_END_TIME.getKey(), "2106-01-08T23:40:53.384Z");
+
+        if (dimensionsFieldLimit > 0) {
```
maybe make `dimensionsFieldLimit` `java.lang.Integer` and have a test which passes dow null?

martijnvg reviewed (approved):
Left one minor comment otherwise LGTM

salvatore-campagna committed:
fix: use Integer instead of int

salvatore-campagna committed:
fix: use null instead of 0

salvatore-campagna committed:
Merge branch 'main' into feature/increase-dimensions-limit
